tools/walletextension/keymanager/keymanager.go

```diff
@@ -0,0 +1,327 @@
+package keymanager
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path/filepath"
+
+	"github.com/edgelesssys/ego/enclave"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	gethlog "github.com/ethereum/go-ethereum/log"
+	tencommon "github.com/ten-protocol/go-ten/go/common"
+	"github.com/ten-protocol/go-ten/go/common/log"
+	"github.com/ten-protocol/go-ten/go/enclave/core/egoutils"
+	"github.com/ten-protocol/go-ten/tools/walletextension/common"
+)
+
+const (
+	dataDir    = "/data"
+	RSAKeySize = 2048
+)
+
+var encryptionKeyFile = filepath.Join(dataDir, "encryption-key.json")
+
+// KeyExchangeRequest represents the structure of the data sent from REQ to OG
+type KeyExchangeRequest struct {
+	PublicKey   []byte `json:"public_key"`
+	Attestation []byte `json:"attestation"`
+}
+
+// KeyExchangeResponse represents the structure of the data sent from OG to REQ
+type KeyExchangeResponse struct {
+	EncryptedKey string `json:"encrypted_key"` // Base64 encoded encrypted encryption key
+}
+
+// GetEncryptionKey returns encryption key for the database
+// 1.) If we use sqlite database we don't need to do anything since sqlite does not need encryption and is usually running in dev environments / for testing
+// 2.) We need to check if the key is already sealed and unseal it if so
+// 3.) If there is a URL to exchange the key with another enclave we need to get the key from there and seal it on this enclave
+// III) If the key is not sealed and we don't have the URL to exchange the key we need to generate a new one and seal it
+func GetEncryptionKey(config common.Config, logger gethlog.Logger) ([]byte, error) {
+	// 1.) check if we are using sqlite database and no encryption key needed
+	if config.DBType == "sqlite" {
+		logger.Info("using sqlite database, no encryption key needed - exiting key exchange process")
+		return nil, nil
+	}
+
+	var encryptionKey []byte
+
+	// 2.) Check if we have a sealed encryption key and try to unseal it
+	encryptionKey, found, err := tryUnsealKey(encryptionKeyFile, config.InsideEnclave)
+	if err != nil {
+		logger.Info("unable to unseal encryption key", log.ErrKey, err)
+	}
+	// If we found a sealed key we can return it
+	if found {
+		logger.Info("found sealed encryption key")
+		return encryptionKey, nil
+	}
+
+	// 3.) We have to exchange the key with another enclave if we have a key exchange url set
+	if config.KeyExchangeURL != "" {
+		encryptionKey, err = HandleKeyExchange(config, logger)
+		if err != nil {
+			logger.Warn("unable to exchange key", log.ErrKey, err)
```

Comment:
I think we have to return the error here. If Gateway has been configured with a key exchange url and something's gone wrong then I don't think it's safe to assume we can create a new key - does that make sense or am I confused?

Reply:
Agreed, changed to `logger.Crit`

Reply:
Also renames OG, REQ to: `KeyProvider` and  `KeyRequester` which are more self/explanatory